### PR TITLE
[Fix] Let a script go three seconds after npm exits when a descendant still holds its output

### PR DESCRIPTION
**Pull request by juanmaguitar**
## Why

A script's run settles on the runner's `close` event, which waits for the stdio pipes. A descendant that outlives npm keeps them open: on Windows, `wp-build` after the console window it ran in was closed (#497). npm exits, the checklist stays **IN PROGRESS**, the button stays busy and the terminal keeps receiving the orphan's lines for as long as it runs (#498). Ctrl+C was the only way out. The `close`-vs-`exit` logic is shared, so a POSIX orphan holding the pipe hangs the same way.

## What changes

`runNpmWithEngineRetry` in `src/main.js`, which installs and scripts share, mirrors the Stop escalation of #479 on the natural-exit path. On the runner's `exit` it arms a three-second grace; `close` stands it down. If the grace runs out it logs why, forces the group by pid (`killTreeByPid`, the same call Stop escalates with) and destroys the runner's stdio pipes. Destroying the pipes is what makes Node emit `close` with the exit code npm gave, so the existing `close` handler settles the run and the engines-retry decision stays in one place. Whatever the orphan writes afterwards goes with the destroyed stream.

On POSIX the detached process group outlives its leader, so the forced signal ends the orphan. On Windows nothing is forced, for the reason `npm:kill` already gives: the pid is certainly dead, a `taskkill /T` on it reaches nothing of the orphan and could land on a reissued pid. There the step ends and the orphan runs on until it finishes on its own. A line in the terminal says which of the two happened.

The Playground and web servers keep their `close`-only lifecycle: their runners load the CLI in-process and have no descendant to orphan.

## How to test this

Platform: either; the mechanism is covered on macOS by an integration test and the Windows case is the one the issue reports. Build: the Buildkite artifact for the current head, or `npm start` from the repository root.

**Windows, Gutenberg site past install:** click **Run full build** and, while it runs, end the `cmd.exe` that runs `wp-build` from Task Manager or Process Explorer (with #497 in, there is no window to close, so break the tree by hand). Expected within about three seconds of npm's `exit 1` in the terminal: the build step turns **FAILED**, the button frees, the terminal stops receiving `wp-build`'s lines. `wp-build` itself may run on in Task Manager until it finishes; that is the stated limitation.

**macOS, by script, on the current head:** a scratch Gutenberg copy with its `build` script replaced by `sh -c 'sleep 90 & echo npm-side-done; exit 1'`; **Run full build**. Expected: the terminal shows `npm-side-done` and npm's exit, the button frees about three seconds later, the terminal says npm exited but something it started was still holding its output, and `pgrep -f 'sleep 90'` finds nothing.

Done on macOS, current head, by a Playwright script driving the app against a scratch copy of a built Gutenberg trunk, the command sent through the app's terminal: npm's side done at 0.5 s with the orphan alive, the terminal freed 2.8 s after npm's exit, the orphan gone 1.5 s later.

**What must not have happened:** a run that closed on its own being forced (the grace is cleared by `close`); a done event sent twice; an install being retried by the let-go (the retry decision is untouched and reads npm's real code).

Tests: `tests/unit/ipc-wiring.test.cjs`, three `#498` cases (script let go after the grace with npm's code; a close inside the grace is left alone; an install the same way), fail on trunk and pass here. `tests/unit/orphan-pipe.integration.test.cjs` (POSIX) pins the Node and POSIX facts against a real orphan: exit without close while it holds the pipes, close with the runner's code once they are destroyed, the orphan ending on the group signal by pid. Lint clean, 1618 unit tests pass.

## Risks and limitations

- Three seconds is the same grace Stop gives; a runner whose pipes take longer than that to drain after exit would be cut short. Nothing prints for three seconds after its own exit.
- On Windows the orphan survives (above). Ending it would need a job object, which Node does not expose.
- The quit sweep and the pre-delete stop are unchanged (#479's own follow-ups).

## Related

Fixes #498. Stacked on #497's PR. Builds on #479. Part of #251, v1.2.0-beta.2.

---

<details>
<summary>Review outcome (required — see AGENTS.md)</summary>

3 [fix here] · 2 [follow-up] on `c6f152c`, the three fixed and one follow-up taken in `f49c77f`.

- Fixed: `killTreeByPid` ran on Windows against a pid that is certainly dead, which `npm:kill`'s own rule forbids; POSIX only now, and the tests assert both branches the way the #479 test does.
- Fixed: the let-go reached only the log file; the terminal now says npm exited with something still holding its output, and whether it was ended or runs on.
- Fixed: the integration test's cleanup went through the pid fallback; it signals the group and the orphan directly, like its sibling.
- Taken: `TESTING.md` layer 2 now names real processes beside real Git repositories.
- Follow-up, not taken: `fakeChild` does not set `exitCode` when a test emits `exit`; harmless while the let-go kills by pid, worth doing if it ever goes through `killChildTree`.
- The review also raised whether a test that spends a second proving a negative belongs in the default suite; one second here, left as is.
</details>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by juanmaguitar** [comment]
@coderabbitai review

File: src/main.js [hunk 1]
```diff
@@ -3428,7 +3428,37 @@ function runNpmWithEngineRetry({ runnerPath, args, cwd, onLog, onDone, register,
...
+		// is gone) and could land on a reissued pid; there the orphan runs on
+		// until it finishes, and the terminal line says so.
+		let letGo = null;
+		child.once('exit', (code, signal) => {
+			letGo = setTimeout(() => {
+				if (settled) return;
+				const exit = `code ${code}${signal ? ` (signal ${signal})` : ''}`;
+				logEvent(logScope, `exited with ${exit} but a descendant still holds its output; letting go`);
+				onLog('stderr', `\nnpm exited with ${exit}, but something it started is still running and holding its output. Letting go${process.platform === 'win32' ? '; that process runs on until it finishes on its own' : ' and ending it'}.\n`);
+				if (process.platform !== 'win32') killTreeByPid(child.pid, 'SIGKILL');
```
[the remark this answers is not in the document]
Comment (juanmaguitar) [reply src/main.js]: Done in b81a422: `killTreeByPid` takes `groupOnly`, and the let-go passes it, so a group that is gone leaves the bare pid alone. The Stop escalation keeps the fallback, for a child that never led a group. Unit test in `kill-tree.test.cjs` covers both.

File: tests/unit/orphan-pipe.integration.test.cjs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	});
+	// Straight to the group, never the bare pid: once the body has reaped the
+	// group, the runner's pid may be someone else's.
+	let orphanPid = null;
+	t.after(() => {
+		try { process.kill(-runner.pid, 'SIGKILL'); } catch {}
+		if (orphanPid !== null) { try { process.kill(orphanPid, 'SIGKILL'); } catch {} }
+	});
+
+	runner.stdout.on('data', (data) => { if (orphanPid === null) orphanPid = Number(String(data).trim()); });
```
[the remark this answers is not in the document]
Comment (juanmaguitar) [reply tests/unit/orphan-pipe.integration.test.cjs]: Done in b81a422, as suggested: the pid is read from the first complete line.
